django/db/models/query.py

```diff
@@ -815,6 +815,25 @@ def complex_filter(self, filter_obj):
         else:
             return self._filter_or_exclude(None, **filter_obj)
 
+    def _combinator_query(self, combinator, *other_qs):
+        # Clone the query so we inherit the select list and everything
+        clone = self._clone()
+        # Clear limits and ordering, so they can be reapplied
+        clone.query.clear_ordering(True)
+        clone.query.clear_limits()
+        clone.query.combined_queries = (self.query,) + tuple(qs.query for qs in other_qs)
+        clone.query.combinator = combinator
+        return CombinedQuerySet(clone)
+
+    def union(self, *other_qs):
```

Comment:
What about accepting a `all=False` kwarg here instead of relying on the `distinct()` method to make the distinction? Else we'll have to make sure not to allow `*expressions` passing to distinct (for `DISTINCT ON`) and limit us if we want to allow `DISTINCT ON` support in the long run.

For example, I could see the following query being useful:

```
SELECT DISTINCT ON field, ... FROM (
    SELECT field, ... FROM t1 WHERE ...
    UNION ALL
    SELECT field, ... FROM t2 WHERE ...
)
```